File: backend-cs/Services/DriveMonitorService.cs
using DriveChill.Models;
using Microsoft.Data.Sqlite;

namespace DriveChill.Services;

/// <summary>
/// Discovers and monitors local drives via an <see cref="IDriveProvider"/>.
/// Publishes hdd_temp sensor readings into SensorService.
/// Polling schedule mirrors the Python backend.
/// </summary>
public sealed class DriveMonitorService : IDisposable
{
    private readonly IDriveProvider _provider;
    private readonly AppSettings _settings;
    private readonly ILogger<DriveMonitorService> _log;
    private readonly SensorService? _sensorService;
    private readonly SmartTrendService? _smartTrend;
    private readonly AlertService? _alertService;
    private DriveHealthNormalizer? _normalizer;

    private readonly Dictionary<string, DriveRawData> _drives = new();
    private readonly object _drivesLock = new();

    private CancellationTokenSource? _cts;
    private Task? _tempTask;
    private Task? _healthTask;
    private Task? _rescanTask;

    private bool _providerAvailable;
    private string _connStr = "";

    public DriveMonitorService(
        IDriveProvider provider,
        AppSettings settings,
        ILogger<DriveMonitorService> log,
        SensorService sensorService,
        SmartTrendService? smartTrend = null,
        AlertService? alertService = null)
    {
        _provider      = provider;
        _settings      = settings;
        _log           = log;
        _sensorService = sensorService;
        _smartTrend    = smartTrend;
        _alertService  = alertService;
    }

    // ── Public accessors ──────────────────────────────────────────────────────

    public IReadOnlyList<DriveRawData> GetAllDrives()
    {
        lock (_drivesLock) return [.. _drives.Values];
    }

    public DriveRawData? GetDrive(string driveId)
    {
        lock (_drivesLock) return _drives.GetValueOrDefault(driveId);
    }

    /// <summary>True when the underlying provider (e.g. smartctl) is available.</summary>
    public bool SmartctlAvailable => _providerAvailable;

    // ── Startup / shutdown ────────────────────────────────────────────────────

    public async Task StartAsync(DriveSettings s, CancellationToken ct = default)
    {
        if (!s.Enabled) return;

        _connStr           = $"Data Source={_settings.DbPath}";
        _normalizer        = new DriveHealthNormalizer(s);
        _providerAvailable = await _provider.CheckAvailableAsync(s, ct);

        _cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        var token = _cts.Token;

        // Initial scan (blocking — must complete before loops start)
        try { await RescanAsync(s, token); }
        catch (Exception ex) { _log.LogWarning(ex, "Drive initial scan failed"); }

        _tempTask   = RunLoop(() => PollTempsAsync(s, token),   s.FastPollSeconds,   token);
        _healthTask = RunLoop(() => PollHealthAsync(s, token),  s.HealthPollSeconds, token);
        _rescanTask = RunLoop(() => RescanAsync(s, token),      s.RescanPollSeconds, token);
    }

    public async Task StopAsync()
    {
        _cts?.Cancel();
        var tasks = new[] { _tempTask, _healthTask, _rescanTask }
            .Where(t => t is not null).Cast<Task>();
        try { await Task.WhenAll(tasks); }
        catch (OperationCanceledException) { }
        catch (AggregateException aex) when (aex.InnerExceptions.All(e => e is OperationCanceledException)) { }
    }

    public void Dispose() { _cts?.Dispose(); }

    // ── Polling loops ─────────────────────────────────────────────────────────

    private async Task RunLoop(Func<Task> work, int intervalSeconds, CancellationToken ct)
    {
        while (!ct.IsCancellationRequested)
        {
            try { await work(); }
            catch (OperationCanceledException) { throw; }
            catch (Exception ex) { _log.LogWarning(ex, "Drive monitor loop iteration failed"); }
            await Task.Delay(TimeSpan.FromSeconds(intervalSeconds), ct);
        }
    }

    // ── Scan / poll implementations ───────────────────────────────────────────

    private async Task RescanAsync(DriveSettings s, CancellationToken ct)
    {
        _providerAvailable = await _provider.CheckAvailableAsync(s, ct);
        if (!_providerAvailable) return;

        var discovered = await _provider.DiscoverDrivesAsync(s, ct);
        var newDrives  = discovered.ToDictionary(d => d.Id);

        lock (_drivesLock)
        {
            var toRemove = _drives.Keys.Except(newDrives.Keys).ToList();
            foreach (var id in toRemove) _drives.Remove(id);
            foreach (var (id, raw) in newDrives)
                _drives[id] = raw;
        }

        await UpsertDrivesAsync(newDrives.Values, ct);
        await PublishDriveSensorsAsync(s, ct);
        _log.LogDebug("Drive rescan complete: {Count} drives", newDrives.Count);
    }

    private async Task PollTempsAsync(DriveSettings s, CancellationToken ct)
    {
        bool updated = false;
        List<DriveRawData> snapshot;
        lock (_drivesLock) snapshot = [.. _drives.Values];

        foreach (var raw in snapshot)
        {
            var refreshed = await _provider.GetDriveDataAsync(raw.DevicePath, s, ct);
            if (refreshed is not null)
            {
                lock (_drivesLock) _drives[raw.Id] = refreshed;
                if (refreshed.TemperatureC.HasValue) updated = true;
            }
        }
        if (updated) await PublishDriveSensorsAsync(s, ct);
    }

    private async Task PollHealthAsync(DriveSettings s, CancellationToken ct)
    {
        List<DriveRawData> snapshot;
        lock (_drivesLock) snapshot = [.. _drives.Values];

        foreach (var raw in snapshot)
        {
            var refreshed = await _provider.GetDriveDataAsync(raw.DevicePath, s, ct);
            if (refreshed is null) continue;

            lock (_drivesLock) _drives[raw.Id] = refreshed;
            if (_normalizer is not null)
                await InsertHealthSnapshotAsync(refreshed, _normalizer, ct);

            // SMART trend detection — inject any new alerts into AlertService
            if (_smartTrend is not null)
            {
                var trendAlerts = _smartTrend.CheckDrive(
                    raw.Id, refreshed.Name,
                    refreshed.ReallocatedSectors,
                    refreshed.WearPercentUsed,
                    refreshed.PowerOnHours);

                if (_alertService is not null)
                {
                    foreach (var ta in trendAlerts)
                    {
                        var syntheticRuleId = $"smart_{raw.Id}_{ta.Condition}";
                        _alertService.InjectEvent(
                            ruleId:      syntheticRuleId,
                            sensorId:    $"hdd_temp_{raw.Id}",
                            sensorName:  refreshed.Name,
                            actualValue: ta.ActualValue,
                            threshold:   ta.Threshold,
                            condition:   "above",
                            message:     ta.Message);
                    }
                }
            }
        }
        await PublishDriveSensorsAsync(s, ct);
    }

    // ── Sensor injection ──────────────────────────────────────────────────────

    private Task PublishDriveSensorsAsync(DriveSettings s, CancellationToken _)
    {
        if (_sensorService is null || _normalizer is null) return Task.CompletedTask;
        List<DriveRawData> drives;
        lock (_drivesLock) drives = [.. _drives.Values];

        var readings = drives
            .Where(d => d.TemperatureC.HasValue)
            .Select(d => new SensorReading
            {
                Id         = $"hdd_temp_{d.Id}",
                Name       = d.Name,
                SensorType = "hdd_temp",
                Value      = d.TemperatureC!.Value,
                MinValue   = 0,
                MaxValue   = _normalizer.TempCriticalC(d) + 20.0,
                Unit       = "°C",
                DriveId    = d.Id,
                EntityName = d.Name,
                SourceKind = d.Capabilities.TemperatureSource == "smartctl" ? "smartctl" : "native",
            })
            .ToList();

        _sensorService.UpdateDriveReadings(readings);
        return Task.CompletedTask;
    }

    // ── Public manual actions ─────────────────────────────────────────────────

    public async Task<int> RescanNowAsync(DriveSettings s, CancellationToken ct = default)
    {
        await RescanAsync(s, ct);
        lock (_drivesLock) return _drives.Count;
    }

    public async Task<DriveRawData?> RefreshDriveAsync(
        string driveId, DriveSettings s, CancellationToken ct = default)
    {
        DriveRawData? raw;
        lock (_drivesLock) raw = _drives.GetValueOrDefault(driveId);
        if (raw is null) return null;

        var refreshed = await _provider.GetDriveDataAsync(raw.DevicePath, s, ct);
        if (refreshed is not null)
        {
            lock (_drivesLock) _drives[driveId] = refreshed;
            await PublishDriveSensorsAsync(s, ct);
        }
        return refreshed ?? raw;
    }

    public async Task<string?> StartSelfTestAsync(
        string driveId, SelfTestType testType, DriveSettings s, CancellationToken ct = default)
    {
        DriveRawData? raw;
        lock (_drivesLock) raw = _drives.GetValueOrDefault(driveId);
        if (raw is null) return null;

        var typeFlag = testType switch
        {
            SelfTestType.Short      => "short",
            SelfTestType.Extended   => "long",
            SelfTestType.Conveyance => "conveyance",
            _ => "short",
        };
        return await _provider.StartSelfTestAsync(raw.DevicePath, typeFlag, s, ct);
    }

    public async Task<bool> AbortSelfTestAsync(
        string driveId, DriveSettings s, CancellationToken ct = default)
    {
        DriveRawData? raw;
        lock (_drivesLock) raw = _drives.GetValueOrDefault(driveId);
        if (raw is null) return false;
        return await _provider.AbortSelfTestAsync(raw.DevicePath, s, ct);
    }

    // ── DB helpers ────────────────────────────────────────────────────────────

    private async Task UpsertDrivesAsync(
        IEnumerable<DriveRawData> drives, CancellationToken ct)
    {
        await using var conn = new SqliteConnection(_connStr);
        await conn.OpenAsync(ct);
        var now = DateTimeOffset.UtcNow.ToString("o");
        await using var tx = await conn.BeginTransactionAsync(ct);
        foreach (var d in drives)
        {
            await using var cmd = conn.CreateCommand();
            cmd.Transaction = (SqliteTransaction)tx;
            cmd.CommandText = """
                INSERT INTO drives
                    (id, name, model, serial_full, device_path, bus_type, media_type,
                     capacity_bytes, firmware_version, smart_available, native_available,
                     supports_self_test, supports_abort, last_seen_at, last_updated_at)
                VALUES ($id,$name,$model,$serial,$devpath,$bus,$media,
                        $cap,$fw,$smart,$native,$selftest,$abort,$now,$now)
                ON CONFLICT(id) DO UPDATE SET
                    name=excluded.name, model=excluded.model,
                    serial_full=excluded.serial_full,
                    device_path=excluded.device_path, bus_type=excluded.bus_type,
                    media_type=excluded.media_type, capacity_bytes=excluded.capacity_bytes,
                    firmware_version=excluded.firmware_version,
                    smart_available=excluded.smart_available,
                    native_available=excluded.native_available,
                    supports_self_test=excluded.supports_self_test,
                    supports_abort=excluded.supports_abort,
                    last_seen_at=excluded.last_seen_at,
                    last_updated_at=excluded.last_updated_at
                """;
            cmd.Parameters.AddWithValue("$id",       d.Id);
            cmd.Parameters.AddWithValue("$name",     d.Name);
            cmd.Parameters.AddWithValue("$model",    d.Model);
            cmd.Parameters.AddWithValue("$serial",   d.Serial);
            cmd.Parameters.AddWithValue("$devpath",  d.DevicePath);
            cmd.Parameters.AddWithValue("$bus",      d.BusType);
            cmd.Parameters.AddWithValue("$media",    d.MediaType);
            cmd.Parameters.AddWithValue("$cap",      d.CapacityBytes);
            cmd.Parameters.AddWithValue("$fw",       d.FirmwareVersion);
            cmd.Parameters.AddWithValue("$smart",    d.Capabilities.SmartRead ? 1 : 0);
            cmd.Parameters.AddWithValue("$native",   0);
            cmd.Parameters.AddWithValue("$selftest", d.Capabilities.SmartSelfTestShort ? 1 : 0);
            cmd.Parameters.AddWithValue("$abort",    d.Capabilities.SmartSelfTestAbort ? 1 : 0);
            cmd.Parameters.AddWithValue("$now",      now);
            await cmd.ExecuteNonQueryAsync(ct);
        }
        await tx.CommitAsync(ct);
    }

    private async Task InsertHealthSnapshotAsync(
        DriveRawData d, DriveHealthNormalizer normalizer, CancellationToken ct)
    {
        await using var conn = new SqliteConnection(_connStr);
        await conn.OpenAsync(ct);
        await using var cmd = conn.CreateCommand();
        cmd.CommandText = """
            INSERT INTO drive_health_snapshots
                (drive_id, recorded_at, temperature_c, health_status, health_percent,
                 predicted_failure, wear_percent_used, available_spare_percent,
                 reallocated_sectors, pending_sectors, uncorrectable_errors,
                 media_errors, power_on_hours, unsafe_shutdowns)
            VALUES ($did,$now,$temp,$health,$healthpct,$pf,$wear,$spare,
                    $realloc,$pending,$uncorr,$media,$poh,$unsafe)
            """;
        var now = DateTimeOffset.UtcNow.ToString("o");
        cmd.Parameters.AddWithValue("$did",       d.Id);
        cmd.Parameters.AddWithValue("$now",       now);
        cmd.Parameters.AddWithValue("$temp",      (object?)d.TemperatureC ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$health",    normalizer.HealthStatus(d));
        cmd.Parameters.AddWithValue("$healthpct", (object?)normalizer.HealthPercent(d) ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$pf",        d.PredictedFailure ? 1 : 0);
        cmd.Parameters.AddWithValue("$wear",      (object?)d.WearPercentUsed ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$spare",     (object?)d.AvailableSparePercent ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$realloc",   (object?)d.ReallocatedSectors ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$pending",   (object?)d.PendingSectors ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$uncorr",    (object?)d.UncorrectableErrors ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$media",     (object?)d.MediaErrors ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$poh",       (object?)d.PowerOnHours ?? DBNull.Value);
        cmd.Parameters.AddWithValue("$unsafe",    (object?)d.UnsafeShutdowns ?? DBNull.Value);
        await cmd.ExecuteNonQueryAsync(ct);
    }

    // ── Helper to build summary/detail shapes ─────────────────────────────────

    public DriveSummary ToSummary(DriveRawData d)
    {
        var health    = _normalizer?.HealthStatus(d) ?? "unknown";
        var healthPct = _normalizer?.HealthPercent(d);
        return new DriveSummary
        {
            Id = d.Id, Name = d.Name, Model = d.Model,
            SerialMasked     = MaskSerial(d.Serial),
            DevicePathMasked = MaskDevicePath(d.DevicePath),
            BusType = d.BusType, MediaType = d.MediaType,
            CapacityBytes    = d.CapacityBytes,
            TemperatureC     = d.TemperatureC,
            HealthStatus     = health, HealthPercent = healthPct,
            SmartAvailable   = d.Capabilities.SmartRead,
            NativeAvailable  = d.Capabilities.HealthSource != "none",
            SupportsSelfTest = d.Capabilities.SmartSelfTestShort,
            SupportsAbort    = d.Capabilities.SmartSelfTestAbort,
        };
    }

    public DriveDetail ToDetail(DriveRawData d, DriveSelfTestRun? lastTest = null)
    {
        var summary = ToSummary(d);
        var warnC = _normalizer?.TempWarningC(d)  ?? 45.0;
        var critC = _normalizer?.TempCriticalC(d) ?? 50.0;
        return new DriveDetail
        {
            Id = summary.Id, Name = summary.Name, Model = summary.Model,
            SerialMasked     = summary.SerialMasked,
            DevicePathMasked = summary.DevicePathMasked,
            BusType = summary.BusType, MediaType = summary.MediaType,
            CapacityBytes    = summary.CapacityBytes,
            TemperatureC     = summary.TemperatureC,
            HealthStatus     = summary.HealthStatus, HealthPercent = summary.HealthPercent,
            SmartAvailable   = summary.SmartAvailable,
            NativeAvailable  = summary.NativeAvailable,
            SupportsSelfTest = summary.SupportsSelfTest,
            SupportsAbort    = summary.SupportsAbort,
            SerialFull            = d.Serial, DevicePath = d.DevicePath,
            FirmwareVersion       = d.FirmwareVersion,
            InterfaceSpeed        = d.InterfaceSpeed,
            RotationRateRpm       = d.RotationRateRpm,
            PowerOnHours          = d.PowerOnHours,
            PowerCycleCount       = d.PowerCycleCount,
            UnsafeShutdowns       = d.UnsafeShutdowns,
            WearPercentUsed       = d.WearPercentUsed,
            AvailableSparePercent = d.AvailableSparePercent,
            ReallocatedSectors    = d.ReallocatedSectors,
            PendingSectors        = d.PendingSectors,
            UncorrectableErrors   = d.UncorrectableErrors,
            MediaErrors           = d.MediaErrors,
            PredictedFailure      = d.PredictedFailure,
            TemperatureWarningC   = warnC,
            TemperatureCriticalC  = critC,
            Capabilities  = d.Capabilities,
            LastSelfTest  = lastTest,
            RawAttributes = d.RawAttributes,
        };
    }

    private static string MaskSerial(string serial) =>
        string.IsNullOrEmpty(serial) ? "****" : "****" + serial[^Math.Min(4, serial.Length)..];

    private static string MaskDevicePath(string path)
    {
        var idx = path.LastIndexOf('/');
        return idx >= 0 ? path[(idx + 1)..] : path;
    }
}

File: backend/app/services/drive_monitor_service.py
"""Drive monitoring service: polling, sensor injection."""
from __future__ import annotations

import asyncio
import logging
from typing import TYPE_CHECKING

import aiosqlite

from app.db.repositories.drive_repo import DriveRepo
from app.db.repositories.settings_repo import SettingsRepo
from app.hardware.drives.composite_provider import CompositeDriveProvider
from app.hardware.drives.native_provider import NativeDriveProvider
from app.hardware.drives.smartctl_provider import SmartctlProvider
from app.models.drives import (
    DriveRawData,
    DriveSettings,
    HealthStatus,
)
from app.models.sensors import SensorReading, SensorType
from app.services.drive_health_normalizer import DriveHealthNormalizer

if TYPE_CHECKING:
    from app.services.sensor_service import SensorService
    from app.services.smart_trend_service import SmartTrendService

logger = logging.getLogger(__name__)


class DriveMonitorService:
    """
    Coordinates drive discovery, polling, health normalization, and
    hdd_temp sensor injection into the main sensor pipeline.

    Polling schedule:
    - Rescan: default 900s — full inventory refresh
    - Health poll: default 300s — health, wear, counters
    - Temp poll: default 15s — temperature only (fast path)

    Drive temperatures are published as hdd_temp SensorReading objects
    and injected into SensorService so they ride the existing sensor/fan-curve
    / WebSocket / alert pipeline without any additional infrastructure.
    """

    def __init__(
        self,
        db: aiosqlite.Connection,
        settings_repo: SettingsRepo,
    ) -> None:
        self._db = db
        self._settings_repo = settings_repo
        self._repo = DriveRepo(db)

        # Built after loading settings
        self._provider: CompositeDriveProvider | None = None
        self._normalizer: DriveHealthNormalizer | None = None

        # In-memory drive cache: drive_id → latest DriveRawData
        self._drives: dict[str, DriveRawData] = {}

        self._sensor_service: SensorService | None = None
        self._smart_trend_svc: SmartTrendService | None = None

        self._running = False
        self._tasks: list[asyncio.Task] = []

    # ── Wiring ──────────────────────────────────────────────────────────────

    def set_sensor_service(self, svc: SensorService) -> None:
        self._sensor_service = svc

    def set_smart_trend_service(self, svc: SmartTrendService) -> None:
        self._smart_trend_svc = svc

    # ── Settings loading ─────────────────────────────────────────────────────

    async def _load_settings(self) -> DriveSettings:
        r = self._settings_repo
        return DriveSettings(
            enabled=bool(int(await r.get("drive_monitoring_enabled") or "1")),
            native_provider_enabled=bool(int(await r.get("drive_native_provider_enabled") or "1")),
            smartctl_provider_enabled=bool(int(await r.get("drive_smartctl_provider_enabled") or "1")),
            smartctl_path=await r.get("drive_smartctl_path") or "smartctl",
            fast_poll_seconds=int(await r.get("drive_fast_poll_seconds") or "15"),
            health_poll_seconds=int(await r.get("drive_health_poll_seconds") or "300"),
            rescan_poll_seconds=int(await r.get("drive_rescan_poll_seconds") or "900"),
            hdd_temp_warning_c=float(await r.get("drive_hdd_temp_warning_c") or "45"),
            hdd_temp_critical_c=float(await r.get("drive_hdd_temp_critical_c") or "50"),
            ssd_temp_warning_c=float(await r.get("drive_ssd_temp_warning_c") or "55"),
            ssd_temp_critical_c=float(await r.get("drive_ssd_temp_critical_c") or "65"),
            nvme_temp_warning_c=float(await r.get("drive_nvme_temp_warning_c") or "65"),
            nvme_temp_critical_c=float(await r.get("drive_nvme_temp_critical_c") or "75"),
            wear_warning_percent_used=float(await r.get("drive_wear_warning_percent_used") or "80"),
            wear_critical_percent_used=float(await r.get("drive_wear_critical_percent_used") or "90"),
        )

    # ── Startup / shutdown ───────────────────────────────────────────────────

    async def start(self) -> None:
        settings = await self._load_settings()
        if not settings.enabled:
            logger.info("Drive monitoring disabled by settings")
            return

        native = NativeDriveProvider()
        smartctl = SmartctlProvider(smartctl_path=settings.smartctl_path)
        self._provider = CompositeDriveProvider(
            native=native,
            smartctl=smartctl,
            prefer_smartctl=settings.smartctl_provider_enabled,
        )
        self._normalizer = DriveHealthNormalizer(settings)

        self._running = True

        # Initial inventory scan before starting loops (blocking is fine here)
        try:
            await self._rescan(settings)
        except Exception:
            logger.exception("Drive initial scan failed")

        # Restart recovery: reconcile any self-tests still marked 'running'
        reconcile_task = asyncio.create_task(self._reconcile_running_tests())

        # Polling loops
        self._tasks = [reconcile_task,
            asyncio.create_task(self._temp_loop(settings)),
            asyncio.create_task(self._health_loop(settings)),
            asyncio.create_task(self._rescan_loop(settings)),
        ]

    async def stop(self) -> None:
        self._running = False
        for t in self._tasks:
            t.cancel()
        for t in self._tasks:
            try:
                await t
            except asyncio.CancelledError:
                pass
        self._tasks.clear()

    # ── Drive state accessors ────────────────────────────────────────────────

    def get_all_drives(self) -> list[DriveRawData]:
        return list(self._drives.values())

    def get_drive(self, drive_id: str) -> DriveRawData | None:
        return self._drives.get(drive_id)

    async def get_settings(self) -> DriveSettings:
        return await self._load_settings()

    async def is_smartctl_available(self) -> bool:
        if self._provider is None:
            return False
        return await self._provider.smartctl_available()

    # ── Polling loops ────────────────────────────────────────────────────────

    async def _temp_loop(self, settings: DriveSettings) -> None:
        while self._running:
            await asyncio.sleep(settings.fast_poll_seconds)
            try:
                await self._poll_temps(settings)
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("Drive temp poll failed")

    async def _health_loop(self, settings: DriveSettings) -> None:
        while self._running:
            await asyncio.sleep(settings.health_poll_seconds)
            try:
                await self._poll_health(settings)
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("Drive health poll failed")

    async def _rescan_loop(self, settings: DriveSettings) -> None:
        while self._running:
            await asyncio.sleep(settings.rescan_poll_seconds)
            try:
                await self._rescan(settings)
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("Drive rescan failed")

    # ── Polling implementations ──────────────────────────────────────────────

    async def _rescan(self, settings: DriveSettings) -> None:
        if self._provider is None:
            return
        drives = await self._provider.discover()
        smartctl_ok = await self._provider.smartctl_available()
        native_ok = await self._provider.native_available()

        for raw in drives:
            self._drives[raw.id] = raw
            await self._repo.upsert_drive(
                id=raw.id,
                name=raw.name,
                model=raw.model,
                serial_full=raw.serial,
                device_path=raw.device_path,
                bus_type=raw.bus_type.value,
                media_type=raw.media_type.value,
                capacity_bytes=raw.capacity_bytes,
                firmware_version=raw.firmware_version,
                smart_available=raw.capabilities.smart_read,
                native_available=native_ok,
                supports_self_test=raw.capabilities.smart_self_test_short,
                supports_abort=raw.capabilities.smart_self_test_abort,
            )
            if raw.raw_attributes:
                await self._repo.upsert_attributes(raw.id, raw.raw_attributes)

        await self._publish_drive_sensors(settings)
        logger.debug("Drive rescan complete: %d drives found", len(drives))

    async def _poll_temps(self, settings: DriveSettings) -> None:
        if self._provider is None:
            return
        updated = False
        for drive_id, raw in list(self._drives.items()):
            refreshed = await self._provider.refresh(raw.device_path)
            if refreshed is not None:
                self._drives[drive_id] = refreshed
                if refreshed.temperature_c is not None:
                    updated = True
        if updated:
            await self._publish_drive_sensors(settings)

    async def _poll_health(self, settings: DriveSettings) -> None:
        if self._provider is None or self._normalizer is None:
            return
        for drive_id, raw in list(self._drives.items()):
            refreshed = await self._provider.refresh(raw.device_path)
            if refreshed is None:
                continue
            self._drives[drive_id] = refreshed

            health = self._normalizer.health_status(refreshed)
            health_pct = self._normalizer.health_percent(refreshed)

            await self._repo.insert_health_snapshot(
                drive_id=drive_id,
                temperature_c=refreshed.temperature_c,
                health_status=health.value,
                health_percent=health_pct,
                predicted_failure=refreshed.predicted_failure,
                wear_percent_used=refreshed.wear_percent_used,
                available_spare_percent=refreshed.available_spare_percent,
                reallocated_sectors=refreshed.reallocated_sectors,
                pending_sectors=refreshed.pending_sectors,
                uncorrectable_errors=refreshed.uncorrectable_errors,
                media_errors=refreshed.media_errors,
                power_on_hours=refreshed.power_on_hours,
                unsafe_shutdowns=refreshed.unsafe_shutdowns,
            )
            if refreshed.raw_attributes:
                await self._repo.upsert_attributes(drive_id, refreshed.raw_attributes)

            # SMART trend detection
            if self._smart_trend_svc:
                self._smart_trend_svc.check_drive(
                    drive_id=drive_id,
                    drive_name=refreshed.name,
                    reallocated_sectors=refreshed.reallocated_sectors,
                    wear_percent_used=refreshed.wear_percent_used,
                    power_on_hours=refreshed.power_on_hours,
                )

        await self._publish_drive_sensors(settings)

    # ── Sensor injection ─────────────────────────────────────────────────────

    async def _publish_drive_sensors(self, settings: DriveSettings) -> None:
        if self._sensor_service is None:
            return
        readings: list[SensorReading] = []
        for raw in self._drives.values():
            if raw.temperature_c is None:
                continue
            warn_c = self._normalizer.temp_warning_c(raw) if self._normalizer else 45.0
            crit_c = self._normalizer.temp_critical_c(raw) if self._normalizer else 50.0
            sensor_id = f"hdd_temp_{raw.id}"
            readings.append(
                SensorReading(
                    id=sensor_id,
                    name=raw.name,
                    sensor_type=SensorType.HDD_TEMP,
                    value=raw.temperature_c,
                    min_value=0.0,
                    max_value=crit_c + 20.0,
                    unit="°C",
                    drive_id=raw.id,
                    entity_name=raw.name,
                    source_kind=(
                        "smartctl" if raw.capabilities.temperature_source.value == "smartctl"
                        else "native"
                    ),
                )
            )
        self._sensor_service.update_drive_readings(readings)

    # ── Manual actions (called from routes) ─────────────────────────────────

    async def rescan_now(self) -> int:
        """Trigger an immediate rescan and return the count of discovered drives."""
        settings = await self._load_settings()
        await self._rescan(settings)
        return len(self._drives)

    async def refresh_drive(self, drive_id: str) -> DriveRawData | None:
        raw = self._drives.get(drive_id)
        if raw is None:
            return None
        if self._provider is None:
            return raw
        refreshed = await self._provider.refresh(raw.device_path)
        if refreshed is not None:
            self._drives[drive_id] = refreshed
            settings = await self._load_settings()
            await self._publish_drive_sensors(settings)
        return refreshed or raw

    # ── Self-test restart recovery ────────────────────────────────────────────

    async def _reconcile_running_tests(self) -> None:
        """On startup, reconcile self-test rows marked 'running' with provider state."""
        if self._provider is None:
            return
        running = await self._repo.get_running_self_tests()
        for row in running:
            drive_id = row["drive_id"]
            run_id = row["id"]
            raw = self._drives.get(drive_id)
            if raw is None:
                await self._repo.update_self_test_run(
                    run_id, status="aborted", failure_message="Drive not found on restart"
                )
                continue
            result = await self._provider.get_self_test_status(raw.device_path)
            if result is None:
                await self._repo.update_self_test_run(
                    run_id, status="aborted", failure_message="Status unavailable on restart"
                )
            elif result.status.value in ("passed", "failed", "aborted"):
                await self._repo.update_self_test_run(
                    run_id,
                    status=result.status.value,
                    progress_percent=result.progress_percent,
                    failure_message=result.failure_message,
                )
            # else still running — leave as-is and the self-test service will poll
